### PR TITLE
Cache cause to sources map

### PR DESCRIPTION
**Pull request by elharo**
@netdpb David, is this the right way to add a calculated field in an autovalue class? Is there any way to make the field final?

### CODE REVIEW

**Comment by netdpb** [comment]
Instead of this, you can add [`@Memoized`](https://github.com/google/auto/blob/master/value/userguide/howto.md#memoize) to the `getCauseToSourceClasses()` method, and AutoValue will do the proper double-checked locking to memoize the value on first call.

**Comment by elharo** [comment]
Sounds good. I added @Memoized. Now I'm wondering if I should use that on all the methods.

File: dashboard/src/main/resources/templates/macros.ftl [hunk 1]
```diff
@@ -8,17 +8,16 @@
...
     <p class="jar-linkage-report">
       ${pluralize(targetClassCount, "target class", "target classes")}
       causing linkage errors referenced from
       ${pluralize(sourceClassCount, "source class", "source classes")}.
     </p>
-    <#list causeToSourceClasses.keySet() as errorCause >
-      <#assign sourceClasses = causeToSourceClasses.get(errorCause) />
-      <p class="jar-linkage-report-cause">${errorCause?html}, referenced from ${
+    <#list jarLinkageReport.getCauses() as cause >
+      <#assign sourceClasses = jarLinkageReport.getCauses(cause) />
```
Comment (netdpb) [inline dashboard/src/main/resources/templates/macros.ftl]: It looks like `getCauses(cause)` should be named `getSourceClasses(cause)`.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/JarLinkageReport.java [hunk 2]
```diff
@@ -88,11 +91,16 @@ public String toString() {
...
   }
-
+  
   /** Returns map from the cause of linkage errors to class names affected by the errors. */
   // TODO this is only used by formatJarLinkageReport in macros.ftl. We should be able to
   // refactor this to make it a lot clearer by removing ImmutableMultimap from the API.
-  public ImmutableMultimap<LinkageErrorCause, String> getCauseToSourceClasses() {
+  @Memoized
+  ImmutableMultimap<LinkageErrorCause, String> getCauseToSourceClasses() {
+    return mapCauseToSourceClasses();
```
Comment (netdpb) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/JarLinkageReport.java]: You can inline the private method now, reverting all lines except the new `@Memoized` line (and maybe keeping the default access).

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/JarLinkageReport.java [hunk 3]
```diff
@@ -36,6 +38,7 @@
  */
 @AutoValue
 public abstract class JarLinkageReport {
+
```
Comment (netdpb) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/JarLinkageReport.java]: Revert this line?

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/JarLinkageReport.java [hunk 4]
```diff
@@ -88,11 +91,16 @@ public String toString() {
     }
     return builder.toString();
   }
-
+
```
Comment (netdpb) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/JarLinkageReport.java]: Revert?

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/JarLinkageReport.java [hunk 5]
```diff
@@ -132,7 +140,20 @@ public String toString() {
   public int getErrorCount() {
     return getCauseToSourceClasses().size();
   }
+  
+  public int getTargetClassCount() {
```
Comment (netdpb) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/JarLinkageReport.java]: I'd recommend Javadoc for all public methods.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageErrorCause.java [hunk 6]
```diff
@@ -20,12 +20,16 @@
...
+// TODO "cause" is unclear. Is it the target or the source? I don't know.
+// Rename things to remove the word "cause" and use more specific terminology.
 abstract class LinkageErrorCause {
 
   /** Returns the reason for the error */
   abstract SymbolNotResolvable.Reason getReason();
 
-  /** Returns the symbol causing the error. It's either class name, field name, or method name. */
+  /** 
+   * Returns the fully package qualified name of the missing or inaccessible symbol.
```
Comment (netdpb) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageErrorCause.java]:
```suggestion
   * Returns the fully qualified name of the missing or inaccessible symbol.
```

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageErrorCause.java [hunk 7]
```diff
@@ -20,12 +20,16 @@
...
+   * This is a class name, field name, or method name. */
```
Comment (netdpb) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageErrorCause.java]: For methods, does this include the parameter types?

File: dashboard/src/main/resources/templates/macros.ftl (hunk 1 above)
Comment (elharo) [reply dashboard/src/main/resources/templates/macros.ftl]: good idea; done

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/JarLinkageReport.java (hunk 4 above)
Comment (elharo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/JarLinkageReport.java]: trailing whitespace

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/JarLinkageReport.java (hunk 5 above)
Comment (elharo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/JarLinkageReport.java]: done

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageErrorCause.java (hunk 6 above)
Comment (elharo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageErrorCause.java]: done

File: dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageErrorCause.java (hunk 7 above)
Comment (elharo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageErrorCause.java]: no, it does not

**Pull request merged**